<a href="https://colab.research.google.com/github/AliAlmasri7/FlyRank-ML/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AliAlmasri7/FlyRank-ML/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

## My Lane: Ranking Signal Analysis

I chose the Ranking Signal Analysis lane because I want to investigate which measurable page signals are most strongly associated with declining search performance. The goal is to identify patterns that can help prioritize pages for review instead of relying only on intuition. This lane combines data analysis with machine learning while keeping the final outcome useful for SEO decision-making. Rather than building a model for its own sake, I want to provide evidence-backed recommendations about which signals appear most important in the available data.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## Research Question

**Decision**

Which pages should be prioritized for SEO investigation based on their observed ranking signals?

**Action**

An SEO specialist reviews the highest-priority pages, updates content, improves on-page optimization, or performs additional analysis before rankings decline further.

**Cost of a Wrong Recommendation**

If the system recommends the wrong pages, valuable time may be spent reviewing healthy pages while pages that actually need attention remain unchanged. This reduces the efficiency of SEO efforts and may lead to missed opportunities for improving search performance.

**Why ML or Data Analysis Helps**

Many ranking signals interact with each other. Instead of relying on a single manually defined rule, data analysis and machine learning can identify combinations of signals that are more strongly associated with declining performance and help prioritize pages more consistently.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [1]:
#Load the dataset
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

import pandas as pd, numpy as np
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# The label: a page is 'declining' when its recent trend is down. Simple, honest starter label.
df["is_declining_label"] = df["trend_direction"].str.lower().eq("down").astype(int)
print(df.shape[0], "pages |  declining rate:", round(df["is_declining_label"].mean(), 3))

30000 pages |  declining rate: 0.542


In [2]:
df.head()

,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct,is_declining_label
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4,1
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7,1
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9,1
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8,0
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7,1


In [3]:
#Pick ranking signals to show the results
signals = [
    "avg_position",
    "ctr",
    "impressions_90d",
    "days_since_last_update"
]

df.groupby("is_declining_label")[signals].mean().round(2)

,avg_position,ctr,impressions_90d,days_since_last_update
is_declining_label,,,,
0,16.82,0.73,5533.31,42.37
1,15.94,0.32,4919.10,49.25


An initial look at the starter dataset shows measurable differences between declining and non-declining pages. Pages labeled as declining have a much lower average CTR (0.32 vs. 0.73), slightly fewer impressions over the last 90 days (4919 vs. 5533), and have gone longer without being updated on average (49.25 vs. 42.37 days). Although the average position is similar (15.94 vs. 16.82), these observed differences suggest that multiple ranking signals may be associated with page decline. This makes Ranking Signal Analysis a suitable direction for investigating which signals are most useful for prioritizing pages for SEO review.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*



This project will identify observed relationships between ranking signals and page performance in the available dataset. The results are intended to support SEO decision-making by highlighting which signals appear to be associated with declining pages and can help prioritize pages for further review.

However, this project will not prove that any ranking signal directly causes changes in search performance. It also will not claim to predict Google's ranking algorithm or guarantee that acting on a recommendation will improve rankings. Any conclusions will be limited to the observed data and should be interpreted as evidence-based, directional, and decision-support rather than causal proof.

## Self-check

- [x] Every section above is filled, including both markdown explanations and supporting code.
- [x] The notebook runs from top to bottom without errors.
- [x] No client names, URLs, or private queries are included.
- [x] My claims use careful language such as observed, measured, directional, and decision-support.
- [x] The notebook has been committed to my repository under `work/notebooks/` and is ready for submission.